File: test/smoke/vmuln/vmuln.c
#include <stdio.h>
#include "assert.h"
#include <unistd.h>

// Select which VMUL[0123] to use
#define VMUL0

#ifdef VMUL0

void vmul(int*a, int*b, int*c, int N){
#pragma omp target map(to: a[0:N],b[0:N]) map(from:c[0:N])
#pragma omp teams distribute parallel for 
   for(int i=0;i<N;i++) {
      c[i]=a[i]*b[i];
   }
}

#elif defined(VMUL1)

void vmul(int*a, int*b, int*c, int N){
#pragma omp target map(to: a[0:N],b[0:N]) map(from:c[0:N])
   for(int i=0;i<N;i++) {
      c[i]=a[i]*b[i];
   }
}

#elif defined(VMUL2)

void vmul(int*a, int*b, int*c, int N){
#pragma omp target map(to: a[0:N],b[0:N]) map(from:c[0:N])
   {
      c[0]=a[0]*b[0];
   }
}

#elif defined(VMUL3)

void vmul(int*a, int*b, int*c, int N){
   int x = a[0];
   int y = b[0];
   int z;
#pragma omp target map(from:z)
   {
      z = x * y;
   }
   c[0] = z;
}

#else
#error no VMULx defined
#endif

int main(){
    int N = 100000;    
    int a[N],b[N],c[N],validate[N];
    int flag=-1; // Mark Success
    for(int i=0;i<N;i++) {
        a[i]=i+1;
        b[i]=i+2;
        validate[i]=a[i]*b[i];
    }

    vmul(a,b,c,N);

#if defined(VMUL2) || defined(VMUL3)
    N = 1;
#endif
    for(int i=0;i<N;i++) {
        if(c[i]!=validate[i]) {
//          print 1st bad index
            if( flag == -1 ) 
              printf("First fail: c[%d](%d) != validate[%d](%d)\n",i,c[i],i,validate[i]);
            flag = i;
        }
    }
    if( flag == -1 ){
        printf("Success\n");
        return 0;
    } else {
        printf("Last fail: c[%d](%d) != validate[%d](%d)\n",flag,c[flag],flag,validate[flag]);
        printf("Fail\n");
        return 1;
    }
}

File: test/smoke/vmuln/vmuln.f90
! Select which VMUL[0123] to use
#define VMUL0

#ifdef VMUL0

subroutine vmul(a, b, c, N)
    implicit none
    integer :: a(N), b(N), c(N)
    integer :: N, i

!$omp target map(to: a,b) map(from: c)
!$omp teams distribute parallel do 
    do i=1,N
        c(i) = a(i) * b(i)
    end do
!$omp end target
end subroutine

#elif defined(VMUL1)

subroutine vmul(a, b, c, N)
    implicit none
    integer :: a(N), b(N), c(N)
    integer :: N, i

!$omp target map(to: a,b) map(from: c)
    do i=1,N
        c(i) = a(i) * b(i)
    end do
!$omp end target
end subroutine

#elif defined(VMUL2)

subroutine vmul(a, b, c, N)
    implicit none
    integer :: a(N), b(N), c(N)
    integer :: N

!$omp target map(to: a,b) map(from: c)
        c(1) = a(1) * b(1)
!$omp end target
end subroutine

#elif defined(VMUL3)

subroutine vmul(a, b, c, N)
    implicit none
    integer :: a(N), b(N), c(N)
    integer :: N, x, y, z
    x = a(1)
    y = b(1)

!$omp target map(from: z)
        z = x * y
!$omp end target
    c(1) = z
end subroutine

#else
#error no VMULx defined
#endif

program vmuldemo
    implicit none
    integer, parameter :: N = 100000
    integer a(N), b(N), c(N), validate(N)
    integer i, num, flag;
    num = N
    
    flag = -1
    do i = 1, N
        a(i) = i+1;
        b(i) = i+2;
        validate(i) = a(i) * b(i);
    enddo

    call vmul(a, b, c, N)

#if defined(VMUL2) || defined(VMUL3)
    num = 1
#endif
    do i = 1, num
        if (c(i) .ne. validate(i)) then
            ! print 1st bad index
            if ( flag .eq. -1 ) then
                write(*, '(A, I0, A, I0, A, I0, A, I0)') "First fail: c(", i, "):", c(i), " != validate(", i, "):", validate(i)
            endif
            flag = i;
        endif
    enddo
    if (flag .eq. -1) then
        write(6,*) "Success"
        call exit(0)
    else
        write(*, '(A, I0, A, I0, A, I0, A, I0)') "Last fail:  c(", flag, "):", c(flag), " != validate(", flag, "):", validate(flag)
        write(6,*) "FAILED"
        call exit(1)
    endif
end program


